lib/std/Thread/Mutex/Recursive.zig

```diff
@@ -0,0 +1,86 @@
+//! A synchronization primitive enforcing atomic access to a shared region of
+//! code known as the "critical section".
+//!
+//! Equivalent to `std.Mutex` except it allows the same thread to obtain the
+//! lock multiple times.
+//!
+//! A recursive mutex is an abstraction layer on top of a regular mutex;
+//! therefore it is recommended to use instead `std.Mutex` unless there is a
+//! specific reason a recursive mutex is warranted.
+
+const std = @import("../../std.zig");
+const Recursive = @This();
+const Mutex = std.Thread.Mutex;
+const assert = std.debug.assert;
+
+mutex: Mutex,
+thread_id: std.Thread.Id,
+lock_count: usize,
+
+pub const init: Recursive = .{
+    .mutex = .{},
+    .thread_id = invalid_thread_id,
+    .lock_count = 0,
+};
+
+/// Acquires the `Mutex` without blocking the caller's thread.
+///
+/// Returns `false` if the calling thread would have to block to acquire it.
+///
+/// Otherwise, returns `true` and the caller should `unlock()` the Mutex to release it.
+pub fn tryLock(r: *Recursive) bool {
+    const current_thread_id = std.Thread.getCurrentId();
+    return tryLockInner(r, current_thread_id);
+}
+
+/// Acquires the `Mutex`, blocking the current thread while the mutex is
+/// already held by another thread.
+///
+/// The `Mutex` can be held multiple times by the same thread.
+///
+/// Once acquired, call `unlock` on the `Mutex` to release it, regardless
+/// of whether the lock was already held by the same thread.
+pub fn lock(r: *Recursive) void {
+    const current_thread_id = std.Thread.getCurrentId();
+    if (!tryLockInner(r, current_thread_id)) {
+        r.mutex.lock();
+        assert(r.lock_count == 0);
+        r.lock_count = 1;
+        @atomicStore(std.Thread.Id, &r.thread_id, current_thread_id, .monotonic);
```

Comment:
Think this could be simplified:
```rs
lock():
  c = LOAD(&count, Unordered)
  if not (c > 0 and LOAD(&current_tid, Unorderd) == tid):
    mutex.lock() // make this failable for tryLock()
    c = LOAD(&count, Unordered)
    assert(c == 0) 
    STORE(&current_id, tid, Unordered)
  STORE(&count, c + 1, Unordered)
 ```

Reply:
TBH, an invalid_thread_id would be simplest:
```rs
lock():
  if LOAD(&current_tid, Unordered) != tid:
    mutex.lock()
    STORE(&current_tid, tid, Unordered)
  count += 1
  
unlock():
  count -= 1
  if count == 0:
    STORE(&current_tid, invalid_tid, Unordered)
    mutex.unlock()
```

Reply:
Nice that last impl looks pretty good.